aff4/volume_group.cc

```diff
@@ -1,123 +1,304 @@
+                return STATUS_OK;
+            }
+        }
+
+        return NOT_FOUND;
     }
 
-    for (int i = 0; i < 256; i++) {
-        std::string urn = aff4_sprintf(
-            "%s%02X", AFF4_IMAGESTREAM_SYMBOLIC_PREFIX, i);
+    AFF4Status VolumeGroup::LocateAndAdd(URN& urn) {
+        if (volume_objs.find(urn) != volume_objs.end()) {
+            // Already loaded....
+            return STATUS_OK;
+        }
+
+        // Check known volume URNs.
+        if (foundVolumes.find(urn) != foundVolumes.end()) {
+
+            std::string volume_to_load = foundVolumes.find(urn)->second;
+            // load and return;
+            AFF4Flusher<FileBackedObject> backing_stream;
+            RETURN_IF_ERROR(NewFileBackedObject(
+                    resolver, volume_to_load,
+                    "read", backing_stream));
+
+            AFF4Flusher<ZipFile> volume;
+            RETURN_IF_ERROR(ZipFile::OpenZipFile(
+                    resolver,
+                    std::move(AFF4Flusher<AFF4Stream>(
+                    backing_stream.release())),
+                    volume));
+
+            AddVolume(std::move(AFF4Flusher<AFF4Volume>(volume.release())));
+            return STATUS_OK;
+        }
+
+        // Search for container.
+        resolver->logger->info("Searching for container {}", urn);
+        for (std::string path : searchPaths) {
+            /* 
+             * Look for all AFF4 in this path, and check it's URN.
+             */
+            ScanForAFF4Volumes(path);
+        }
+        // Check known volume URNs.
+        if (foundVolumes.find(urn) != foundVolumes.end()) {
 
-        if (stream_urn == urn) {
-            result.reset(new AFF4SymbolicStream(resolver, stream_urn, i));
+            std::string volume_to_load = foundVolumes.find(urn)->second;
+            
+            resolver->logger->info("Searching for container {} = {}", urn, volume_to_load);
+            
+            // load and return;
+            AFF4Flusher<FileBackedObject> backing_stream;
+            RETURN_IF_ERROR(NewFileBackedObject(
+                    resolver, volume_to_load,
+                    "read", backing_stream));
+
+            AFF4Flusher<ZipFile> volume;
+            RETURN_IF_ERROR(ZipFile::OpenZipFile(
+                    resolver,
+                    std::move(AFF4Flusher<AFF4Stream>(
+                    backing_stream.release())),
+                    volume));
+
+            AddVolume(std::move(AFF4Flusher<AFF4Volume>(volume.release())));
             return STATUS_OK;
         }
+
+        return NOT_FOUND;
     }
 
-    return NOT_FOUND;
+    bool VolumeGroup::FoundVolumesContains(const std::string& filename) {
+	for (auto it = foundVolumes.begin(); it != foundVolumes.end(); it++) {
+		if (it->second.compare(filename) == 0) {
+			return true;
+		}
+	}
+	return false;
 }
 
+    void VolumeGroup::ScanForAFF4Volumes(const std::string& path) {
+        // It is expected that the map is LOCKED prior to this call.
```

Comment:
There is no locking semantics at all in this library - this expectation is not valid.